### PR TITLE
New Kirby\Toolkit\Date class

### PR DESCRIPTION
**Pull request by bastianallgeier**
## Features

- New `Kirby\Toolkit\Date` class as extension for PHP's DateTime class. This new class helps with all kinds of date and time related tasks and is now used by our date and time fields and the `timestamp()` helper.

## Fixes

- https://github.com/getkirby/kirby/issues/3642
- https://github.com/getkirby/kirby/issues/3106
- https://github.com/getkirby/kirby/issues/3842

## Ready?
- [x] Unit tests for fixed bug/feature
- [x] In-code documentation (wherever needed)
- [x] CI checks pass

## When merging

- [ ] Add to [website docs release checklist](https://github.com/getkirby/getkirby.com/pulls) (if needed)
- [ ] Add changes to release notes draft in Notion

### CODE REVIEW

File: config/helpers.php [hunk 1]
```diff
@@ -785,60 +786,11 @@ function tc(string $key, int $count)
...
-        $parts['year'] . '-' .
-        str_pad($parts['month'], 2, 0, STR_PAD_LEFT) . '-' .
-        str_pad($parts['day'], 2, 0, STR_PAD_LEFT) . ' ' .
-        str_pad($parts['hour'], 2, 0, STR_PAD_LEFT) . ':' .
-        str_pad($parts['minute'], 2, 0, STR_PAD_LEFT) . ':' .
-        str_pad($parts['second'], 2, 0, STR_PAD_LEFT)
-    );
+    $date = Date::optional($date);
+    $step = Date::stepConfig($step);
+    $date->round($step['unit'], $step['size']);
```
Comment (distantnative) [inline config/helpers.php]: A few potential breaking changes:
- doesn't check explicitly anymore if the string is a valid date
- always does rounding, event when no step is passed

File: config/fields/date.php [hunk 2]
```diff
@@ -126,51 +105,44 @@
 
             return $this->time['step'];
```
Comment (distantnative) [inline config/fields/date.php]: I think we have an issue here (also had before) that we never teat the sub options of `time` (e.g. `step`) as we prepare them on the top-level

File: config/fields/date.php [hunk 3]
```diff
@@ -46,42 +46,24 @@
...
-                return array_merge($default, ['size' => $step]);
-            }
-
-            if (is_string($step) === true) {
-                return array_merge($default, ['unit' => strtolower($step)]);
-            }
+            return Date::stepConfig($step, [
+                'size' => 5,
+                'unit' => 'minute',
+            ]);
```
Comment (distantnative) [inline config/fields/date.php]: Default step for the date field should be 1 day. Only when `time` option is active, default step should be 5 minutes.

File: config/fields/date.php (hunk 2 above)
Comment (bastianallgeier) [reply config/fields/date.php]: Sorry, I'm lost here. What do you mean exactly :)

Comment (distantnative) [reply config/fields/date.php]: In this computed property we either use the main step option or if a time option is defined, we use the step sub-option of the time option.

But while we would run the main step option through the Date class method, we just use the step sub-option as is.

Comment (bastianallgeier) [reply config/fields/date.php]: Ah, I think I got it.

**Review by distantnative — APPROVED** [review]

**Pull request merged**